### DISCUSSION TITLE
How to package my custom JavaScript Libraries?

### DISCUSSION THREAD

**Discussion by thienvc** [dnnsoftware/Dnn.Platform#4317] [Q&A]
How to package my custom JavaScript Libraries?

**Comment by bdukes** [comment]
You can review the packages available on https://github.com/EngageSoftware/DNN-JavaScript-Libraries for some ideas.  It's pretty much just a zip file with a manifest and a JS file.

**Comment by thienvc** [comment]
@bdukes 
Thanks for your reply, 
But I see in EngageSoftware there is dnn-library.json, which is different from DNN package structure
I want the directory structure to be the same as DNN_Platform and the Build way similar to DNN_Platform solution?
I still haven't figured out how DNN_Platform Solution can package Javascript Libraries

**Comment by thienvc** [comment]
I searched but still can't find out DNN_Platform called msbuild to package the JS library somewhere

**Reply by bdukes** [reply]
Ah, the Platform uses Cake to package the JS libraries that it includes.  See [`thirdparty.cake`](https://github.com/dnnsoftware/Dnn.Platform/blob/dbeb24c99d9f84b4d4cf69546e7c43b3100f6bc1/Build/Cake/thirdparty.cake) and [`thirdparty.json`](https://github.com/dnnsoftware/Dnn.Platform/blob/dbeb24c99d9f84b4d4cf69546e7c43b3100f6bc1/Build/Cake/thirdparty.json)

**Reply by bdukes** [reply]
Which is to say, there isn't a "standard" way to package these outside of the main DNN Platform solution.  Since they're just a zip file, you have a lot of options.

**Reply by thienvc** [reply]
oh oh, thanks you so much, This is what I was looking for hihi, thanks @bdukes

**Comment by valadas** [comment]
And well, depending on your preferences and knowledge of different build tools. It's also possible just using msbuild scripts. most of the older core modules use that such as this one https://github.com/DNNCommunity/DNN.FormAndList/blob/development/BuildScripts/ModulePackage.targets

Then Cake is certainly a popular option and I personally love Nuke https://nuke.build/

All that to say, it is up to your preference :)  Any type of Dnn extension is simply just a zip file that contains a valid manifest. How it get's in there is up to you.
More info on the dnn manifest can be found here https://dnndocs.com/content/tutorials/extensions/dnn-manifest-schema/index.html#the-dnn-manifest-schema

And then I do recommend testing your zip using the extension verification service by uploading it here https://evs.dotnetnuke.com/ It is really helpful in figuring out extension packaging problems since the Dnn installer is not currently great at giving meaningful extension install failures :)

**Reply by thienvc** [reply]
Thanks @valadas 
I know there are many ways to package DNN Extentions, but I mean I want to go the way Core (DNN Platform) does, to simply upgrade DNN later I have less customization of the development environment.

**Reply by valadas** [reply]
The platform uses a mix of MSBuild scripts and Cake. How you decide to package does not matter for Dnn upgrades, it's just whatever way you prefer to get the files in a zip with it's manifest. You could even just do it manually and the result would be the same. Build scripts just help avoiding that manual step.

**Reply by valadas** [reply]
And it may be a bit overcomplex in the platform because we package many many different extension types, so each project has a partial MSBuild script to define some variables, then there are more global scripts that hook up into those variables for packaging and it get's fired up by cake. So I would say doing it exactly like in the platform might not be a great idea for a single extension.

**Comment by roman-yagodin** [comment]
Here's another project for packaging JavaScript libraries for DNN: https://github.com/roman-yagodin/R7.Dnn.JavaScriptLibraries/
It's almost pure msbuild.
